### PR TITLE
Adding OpenSSL 3 API compatibility to openenclave

### PR DESCRIPTION
**Pull request by justanotherminh**
This pull request introduce changes to the openenclave core library, as well as tests, that ensure compatibility with OpenSSL 3 on both the host and the enclave side. To ensure backward compatibility with OpenSSL 1.1 users, each function is rewritten using new OpenSSL 3 APIs while the original functions using OpenSSL 1.1 APIs are kept intact, with the correct version being chosen by the compiler during compile time. The exception is key.c, which cannot be upgraded to OpenSSL 3 due to ENGINE not being supported by OpenSSL 3.

### CODE REVIEW

**Comment by radhikaj** [comment]
bors try

File: 3rdparty/openssl/include/bn_conf.h
Comment (dcarpente) [inline 3rdparty/openssl/include/bn_conf.h:1]: We shouldn't need any changes to bn_conf.h or dso_conf.h, can you undo those changes?

File: host/CMakeLists.txt [hunk 1]
```diff
@@ -136,6 +136,9 @@ if (UNIX)
   list(APPEND PLATFORM_SDK_ONLY_SRC ${PROJECT_SOURCE_DIR}/common/asn1.c
        ${PROJECT_SOURCE_DIR}/common/crypto/openssl/hmac.c
        crypto/openssl/random.c linux/syscall.c)
+  set_source_files_properties(
```
Comment (dcarpente) [inline host/CMakeLists.txt:139]: Can you add a comment explaining why we need -Wno-deprecated-declarations for common/crypto/openssl/key.c?

File: tests/crypto/ec_tests.c [hunk 2]
```diff
@@ -238,7 +238,6 @@ static void _test_generate_from_private()
         sizeof(private_raw),
         &private_key,
         &public_key);
-    OE_TEST(r == OE_OK);
```
Comment (dcarpente) [inline tests/crypto/ec_tests.c:241]: I don't think we should remove this part of the test

File: tests/openssl/enc/CMakeLists.txt [hunk 3]
```diff
@@ -335,6 +345,10 @@ function (add_openssl_test_enc NAME BUILDTEST CRYPTO_LIB)
     list(APPEND INCLUDE_PATHS ${TEST_OPENSSL_DIR}/crypto/rsa)
   endif ()
 
+  if ("${NAME}" STREQUAL "rsa_sp800_56b_test")
```
Comment (dcarpente) [inline tests/openssl/enc/CMakeLists.txt:348]: This looks like it was added by a merge conflict, we're adding the same rsa_sp800_56b_test twice

File: common/crypto/openssl/cmac.c [hunk 4]
```diff
@@ -44,3 +49,58 @@ oe_result_t oe_aes_cmac_sign(
...
+    oe_aes_cmac_t* aes_cmac)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    size_t key_size_bits = key_size * 8;
+    size_t final_size = sizeof(oe_aes_cmac_t);
+    EVP_MAC_CTX* ctx = NULL;
+    EVP_MAC* mac = NULL;
+    OSSL_PARAM params[3];
+
+    if (aes_cmac == NULL)
```
Comment (mingweishih) [inline common/crypto/openssl/cmac.c:67]: We should check `key` and `message` as well (also for the 1.1.1 implementation).

File: tests/openssl/enc/CMakeLists.txt [hunk 5]
```diff
@@ -29,13 +29,19 @@ add_enclave_library(
   # Ensure that openssl_t.h is generated.
   openssl_t.h)
 
+enclave_compile_definitions(openssl-support PRIVATE __STDC_NO_ATOMICS__)
```
Comment (dcarpente) [inline tests/openssl/enc/CMakeLists.txt:32]: I think all the changes in tests/openssl/enc/CMakeLists.txt might be from merge conflicts, I don't think this file should be changed at all

File: tests/crypto/host/CMakeLists.txt [hunk 6]
```diff
@@ -35,7 +35,6 @@ add_executable(
   ../sha_tests.c
   ../tests.c
   ../utils.c)
-
```
Comment (dcarpente) [inline tests/crypto/host/CMakeLists.txt:38]: Remove this noop change

File: tests/tls_e2e/server_enc/CMakeLists.txt [hunk 7]
```diff
@@ -44,6 +44,6 @@ if (BUILD_OPENSSL)
 
   enclave_include_directories(openssl_tls_server_enc PRIVATE
                               ${CMAKE_CURRENT_BINARY_DIR})
-
+
```
Comment (dcarpente) [inline tests/tls_e2e/server_enc/CMakeLists.txt:47]: Remove this noop change. (This kind of change should get caught by trim trailing whitespace tools e.g. https://stackoverflow.com/questions/30884131/remove-trailing-spaces-automatically-or-with-a-shortcut)

File: host/CMakeLists.txt (hunk 1 above)
Comment (dcarpente) [reply host/CMakeLists.txt:139]: I see you have it in the PR description "which cannot be upgraded to OpenSSL 3 due to ENGINE not being supported by OpenSSL 3." I think it would be nice to have as a comment in this file too.

File: common/crypto/openssl/ec.c [hunk 8]
```diff
@@ -61,7 +62,44 @@ static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
...
+    ctx = OSSL_ENCODER_CTX_new_for_pkey(
+        pkey, EVP_PKEY_KEYPAIR, "PEM", NULL, NULL);
+
+    if (!ctx)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    if (!OSSL_ENCODER_to_data(ctx, &buffer, &bytes_written))
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    if (buffer == NULL || bytes_written <= 0)
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:82]: `size_t` cannot be negative.

File: common/crypto/openssl/ec.c [hunk 9]
```diff
@@ -320,6 +390,128 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+    OSSL_PARAM_BLD* bld = NULL;
+    OSSL_PARAM* params = NULL;
+
+    if (!private_key_buf || !private_key || !public_key ||
+        private_key_buf_size > OE_INT_MAX)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+#if !defined(OE_BUILD_ENCLAVE)
+    /* Initialize OpenSSL. */
+    oe_crypto_initialize();
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:420]: We don't need this for OpenSSL 3. See https://github.com/openenclave/openenclave/blob/master/host/crypto/openssl/init.c#L13

File: common/crypto/openssl/ec.c [hunk 10]
```diff
@@ -320,6 +390,128 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+#endif
+
+    if (!(group = EC_GROUP_new_by_curve_name(_get_nid(curve))))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    /* Allocate point */
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:425]: nit: add an empty line above

**Comment by dcarpente** [comment]
bors try

File: common/crypto/openssl/ec.c [hunk 11]
```diff
@@ -440,6 +633,89 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+    EVP_PKEY_CTX* ctx = NULL;
+    OSSL_PARAM_BLD* param_bld = NULL;
+    OSSL_PARAM* params = NULL;
+
+    if (public_key)
+        oe_secure_zero_fill(public_key, sizeof(oe_ec_public_key_t));
+
+#if !defined(OE_BUILD_ENCLAVE)
+    /* Initialize OpenSSL */
+    oe_crypto_initialize();
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:658]: See comments above

File: tests/openssl/tests.supported.buildtest
Comment (dcarpente) [inline tests/openssl/tests.supported.buildtest:1]: tests.supported.buildtest and tests.unsupported shouldn't need any changes, we should remove these changes

File: common/crypto/openssl/ec.c [hunk 12]
```diff
@@ -440,6 +633,89 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+        OE_RAISE(OE_FAILURE);
+
+    /* Create the public EC key */
+    if (!(param_bld = OSSL_PARAM_BLD_new()))
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    if (!OSSL_PARAM_BLD_push_utf8_string(
+            param_bld, "group", OBJ_nid2sn(nid), 0))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    unsigned char* pub_data = OPENSSL_malloc(1 + x_size + y_size);
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:677]: nit: add an empty line above.

File: common/crypto/openssl/ec.c [hunk 13]
```diff
@@ -440,6 +633,89 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    params = OSSL_PARAM_BLD_to_param(param_bld);
+
+    ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
+    if (ctx == NULL || params == NULL || EVP_PKEY_fromdata_init(ctx) <= 0 ||
+        EVP_PKEY_fromdata(ctx, &pkey, EVP_PKEY_PUBLIC_KEY, params) <= 0)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    OPENSSL_free(pub_data);
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:696]: Move the declaration of `pub_data` to the beginning of the function, and move this call after `done`, similar to other variables that needed to be freed.

File: common/crypto/openssl/ec.c [hunk 14]
```diff
@@ -560,7 +836,7 @@ bool oe_ec_valid_raw_private_key(
     if (bn != NULL)
         BN_clear_free(bn);
     if (group != NULL)
-        EC_GROUP_clear_free(group);
+        EC_GROUP_free(group);
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:839]: Can you check if `EC_GROUP_free` be sufficient?
For 1.1.1 (https://www.openssl.org/docs/man1.1.1/man3/EC_GROUP_clear_free.html):
```
EC_GROUP_free() frees the memory associated with the EC_GROUP. If group is NULL nothing is done.

EC_GROUP_clear_free() destroys any sensitive data held within the EC_GROUP and then frees its memory. If group is NULL nothing is done.
```

Not sure about OpenSSL 3.

File: common/crypto/openssl/hmac.c [hunk 15]
```diff
@@ -67,7 +90,56 @@ oe_result_t oe_hmac_sha256_init(
...
+    const uint8_t* key,
+    size_t keysize)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    EVP_MAC_CTX* ctx = NULL;
+    EVP_MAC* mac = NULL;
+    OSSL_PARAM params[3];
+
+#if defined(OE_BUILD_ENCLAVE)
+    OPENSSL_init_crypto(OPENSSL_INIT_NO_LOAD_CONFIG, NULL);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:105]: see comments above.

File: common/crypto/openssl/hmac.c [hunk 16]
```diff
@@ -67,7 +90,56 @@ oe_result_t oe_hmac_sha256_init(
...
+#endif
 
+    if (!context || !key || keysize > OE_INT_MAX)
+        OE_RAISE(OE_INVALID_PARAMETER);
+    mac = EVP_MAC_fetch(NULL, OSSL_MAC_NAME_HMAC, NULL);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:110]: nit: add an empty line above

File: common/crypto/openssl/hmac.c [hunk 17]
```diff
@@ -112,7 +207,34 @@ oe_result_t oe_hmac_sha256_final(
...
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_hmac_sha256_context_impl_t* impl =
+        (oe_hmac_sha256_context_impl_t*)context;
+    size_t hmac_size = 0;
+
+    if (!context || !sha256)
+        OE_RAISE(OE_INVALID_PARAMETER);
 
+    // EVP_MAC_final(impl->ctx, NULL, &hmac_size, 0);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:223]: Remove commented code

File: common/crypto/openssl/key.c [hunk 18]
```diff
@@ -464,7 +464,9 @@ oe_result_t oe_private_key_sign(
 
     /* Create signing context */
     if (!(ctx = EVP_PKEY_CTX_new(impl->pkey, NULL)))
+    {
```
Comment (mingweishih) [inline common/crypto/openssl/key.c:467]: nit: unnecessary change

File: common/crypto/openssl/rsa.c [hunk 19]
```diff
@@ -52,7 +53,44 @@ static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
...
+    ctx = OSSL_ENCODER_CTX_new_for_pkey(
+        pkey, EVP_PKEY_KEYPAIR, "PEM", NULL, NULL);
+
+    if (!ctx)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    if (!OSSL_ENCODER_to_data(ctx, &buffer, &bytes_written))
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    if (buffer == NULL || bytes_written <= 0)
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:73]: `size_t` cannot be negative

File: common/crypto/openssl/rsa.c [hunk 20]
```diff
@@ -121,6 +159,73 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    /* Check for invalid parameters */
+    if (!public_key || !buffer_size)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* If buffer is null, then buffer_size must be zero */
+    if (!buffer && *buffer_size != 0)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Select modulus or exponent */
+    EVP_PKEY_get_bn_param(public_key->pkey, OSSL_PKEY_PARAM_RSA_N, &n);
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:184]: Can this function fail?

File: host/crypto/openssl/rsa.c [hunk 21]
```diff
@@ -123,7 +124,62 @@ oe_result_t oe_rsa_get_public_key_from_private(
...
+    oe_private_key_t* private_key_temp = (oe_private_key_t*)private_key;
+    EVP_PKEY* rsa_public_pkey = NULL;
+    BIGNUM* public_e = NULL;
+    BIGNUM* public_n = NULL;
+
+    /* Check for invalid parameters */
+    if (!private_key || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // RSA_get0_key(rsa_private, &public_e, &public_n, NULL);
```
Comment (mingweishih) [inline host/crypto/openssl/rsa.c:142]: remove commented code

File: host/crypto/openssl/rsa.c [hunk 22]
```diff
@@ -123,7 +124,62 @@ oe_result_t oe_rsa_get_public_key_from_private(
...
+    EVP_PKEY_get_bn_param(
```
Comment (mingweishih) [inline host/crypto/openssl/rsa.c:143]: Can this function fail?

File: host/crypto/openssl/rsa.c [hunk 23]
```diff
@@ -123,7 +124,62 @@ oe_result_t oe_rsa_get_public_key_from_private(
...
+
+    if (!public_n)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    /* Init the OE public key type. */
+    OSSL_PARAM_BLD* bld = OSSL_PARAM_BLD_new();
+    OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_N, public_n);
+    OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, public_e);
+    OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(bld);
+    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
```
Comment (mingweishih) [inline host/crypto/openssl/rsa.c:159]: Does `ctx` need to be freed?

File: tests/attestation_cert_apis/enc/openssl_rsa.cpp [hunk 24]
```diff
@@ -126,3 +127,117 @@ oe_result_t generate_rsa_pair(
...
+        BIO_free(bio);
+    if (e)
+        BN_free(e);
+    if (ctx)
+        EVP_PKEY_CTX_free(ctx);
+    if (pkey)
+        EVP_PKEY_free(pkey);
+    return result;
+}
+#endif
```
Comment (mingweishih) [inline tests/attestation_cert_apis/enc/openssl_rsa.cpp:243]: newline

**Comment by dcarpente** [comment]
bors try

**Comment by dcarpente** [comment]
bors try-

**Comment by dcarpente** [comment]
bors try

File: common/crypto/openssl/ec.c (hunk 14 above)
Comment (justanotherminh) [reply common/crypto/openssl/ec.c:840]: `EC_GROUP_clear_free()` is deprecated in 3. Its functionality is now part of EC_GROUP_free()

Comment (mingweishih) [reply common/crypto/openssl/ec.c:840]: I get that. But now this change applies to both OpenSSL 1 and 3. My question was
For OpenSSL 1, would using `EC_GROUP_free` be sufficient (without introdcuing memory leak)?
Also, please check if it's sufficient on OpenSSL 3 as well.

Comment (justanotherminh) [reply common/crypto/openssl/ec.c:869]: I see what you mean. I added a OpenSSL 3 version like the rest.

**Comment by radhikaj** [comment]
bors try

File: 3rdparty/openssl/include/bn_conf.h
Comment (dcarpente) [reply 3rdparty/openssl/include/bn_conf.h:1]: @justanotherminh It looks like these files are still in the commit

File: common/crypto/openssl/rsa.c (hunk 20 above)
Comment (mingweishih) [reply common/crypto/openssl/rsa.c:184]: Per https://manpages.debian.org/experimental/libssl-doc/EVP_PKEY_get_bn_param.3ssl.en.html, this function can fail. Please check the return value.

File: common/crypto/openssl/rsa.c [hunk 25]
```diff
@@ -121,6 +159,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    if (!n)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    EVP_PKEY_get_bn_param(public_key->pkey, OSSL_PKEY_PARAM_RSA_E, &e);
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:188]: Same here

File: host/crypto/openssl/rsa.c (hunk 22 above)
Comment (mingweishih) [reply host/crypto/openssl/rsa.c:142]: Check return value

File: host/crypto/openssl/rsa.c [hunk 26]
```diff
@@ -123,7 +124,65 @@ oe_result_t oe_rsa_get_public_key_from_private(
...
+    if (!private_key || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    EVP_PKEY_get_bn_param(
+        private_key_temp->pkey, OSSL_PKEY_PARAM_RSA_N, &public_n);
+
+    if (!public_n)
+        OE_RAISE(OE_CRYPTO_ERROR);
+
+    EVP_PKEY_get_bn_param(
```
Comment (mingweishih) [inline host/crypto/openssl/rsa.c:148]: Same here

**Review by mingweishih — APPROVED** [review]
Great work! Please squash the commits before merging.

**Review by dcarpente — APPROVED** [review]
